### PR TITLE
Master data, Lead Creation

### PR DESCRIPTION
**Pull request by manojkrsaini09**
1. Dynamic master data creation
2. lead creation ( lead contact creation is pending)

### CODE REVIEW

File: src/main/java/creativei/controller/LeadController.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import creativei.vo.ResponseObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping("/lead")
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/LeadController.java]: Should be /leads

Comment (aayushKhandpur) [reply src/main/java/creativei/controller/LeadController.java]: Best practice is to always use plural endpoint.

File: src/main/java/creativei/controller/MasterDataController.java [hunk 2]
```diff
@@ -2,26 +2,71 @@
...
 public class MasterDataController {
 
     @Autowired
     private JpaMasterDataDao jpaMasterDataDao;
 
     @RequestMapping(method = RequestMethod.POST)
     @ResponseBody
-    public ResponseEntity<MasterData> createMAsterData(@RequestBody MasterData masterData) {
-        return  new ResponseEntity<MasterData>(jpaMasterDataDao.save(masterData), HttpStatus.OK);
+    public ResponseEntity<ResponseObject> createMasterData(@RequestBody MasterData masterData) {
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/MasterDataController.java]: Use path variable to pass type. /{type}

File: src/main/java/creativei/controller/MasterDataController.java [hunk 3]
```diff
@@ -2,26 +2,71 @@
...
+       String value = masterData.getValue();
+       MasterDataType type = masterData.getType();
+       boolean exist = false;
+       MasterData masterDataCreated = null;
+        if(type != null && value != null){
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/MasterDataController.java]: Pass type as path variable and mark it required = true. No need to check this then.

File: src/main/java/creativei/controller/MasterDataController.java [hunk 4]
```diff
@@ -2,26 +2,71 @@
...
+            masterDataCreated = jpaMasterDataDao.findByTypeAndValue(type,value);
+            if(masterDataCreated != null && masterDataCreated.getId() > 0){
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/MasterDataController.java]: Use !ObjectUtils.isEmpty and avoid double condition.

File: src/main/java/creativei/controller/MasterDataController.java [hunk 5]
```diff
@@ -2,26 +2,71 @@
...
+                return new ResponseEntity<>(ResponseObject.getResponse(ExceptionType.DATA_ALREDY_EXIST.getMessage()
+                        , ExceptionType.DATA_ALREDY_EXIST.getCode()),HttpStatus.INTERNAL_SERVER_ERROR);
+            }
+        }
+        return  new ResponseEntity<>(ResponseObject.getResponse(jpaMasterDataDao.save(masterData)),HttpStatus.OK);
+    }
+
+    @RequestMapping(method = RequestMethod.POST, value = "/masterdata/saveAll")
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/MasterDataController.java]: no need to use /masterdata/saveall. /masterdata is already used as base url. Use POST /all

File: src/main/resources/public/application.js [hunk 6]
```diff
@@ -1094,6 +1094,120 @@ app.config(['$stateProvider', '$urlRouterProvider', '$locationProvider', '$httpP
...
+                                    });
+
+                                }],
+                               }
+                            })
+                                  // *********************************************************//
+                                  // End Of sales  and  marketing states                      //
+                                  // *********************************************************//
             			;
 		$urlRouterProvider.otherwise('/dashboard');
```
Comment (aayushKhandpur) [inline src/main/resources/public/application.js]: IS this base dashboard url?

Comment (manojkrsaini09) [reply src/main/resources/public/application.js]: yes it is multicomponent view route.

File: src/main/java/creativei/controller/MasterDataController.java (hunk 5 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/controller/MasterDataController.java]: done

File: src/main/java/creativei/controller/MasterDataController.java (hunk 4 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/controller/MasterDataController.java]: done

File: src/main/java/creativei/controller/LeadController.java (hunk 1 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/controller/LeadController.java]: done

File: src/main/java/creativei/controller/MasterDataController.java (hunk 3 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/controller/MasterDataController.java]: done

File: src/main/java/creativei/controller/MasterDataController.java (hunk 2 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/controller/MasterDataController.java]: done

**Comment by manojkrsaini09** [comment]
Committed lead contact creation code.

File: src/main/java/creativei/controller/MasterDataController.java (hunk 4 above)
Comment (aayushKhandpur) [reply src/main/java/creativei/controller/MasterDataController.java]: Why do we need null check on id? Id will never be null from database

File: src/main/java/creativei/controller/MasterDataController.java [hunk 7]
```diff
@@ -2,26 +2,69 @@
...
 public class MasterDataController {
 
     @Autowired
     private JpaMasterDataDao jpaMasterDataDao;
 
     @RequestMapping(method = RequestMethod.POST)
     @ResponseBody
-    public ResponseEntity<MasterData> createMAsterData(@RequestBody MasterData masterData) {
-        return  new ResponseEntity<MasterData>(jpaMasterDataDao.save(masterData), HttpStatus.OK);
+    public ResponseEntity<ResponseObject> createMasterData(@RequestParam(value = "type", required = true)  MasterDataType type, @RequestBody MasterData masterData) {
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/MasterDataController.java]: Should add @RequestMapping(value = /{type}, method = RequestMethod.POST ) for better clarity.

File: src/main/java/creativei/controller/MasterDataController.java [hunk 8]
```diff
@@ -2,26 +2,69 @@
...
+       String value = masterData.getValue();
+       boolean exist = false;
+       MasterData masterDataCreated = null;
+        if(type != null && value != null){
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/MasterDataController.java]: Remove this piece of code. type will never be null. Why do we need value now?

**Comment by manojkrsaini09** [comment]
added lead detail page

File: src/main/java/creativei/controller/MasterDataController.java (hunk 4 above)
Comment (manojkrsaini09) [reply src/main/java/creativei/controller/MasterDataController.java]: removed it. It is earlier code.

File: src/main/java/creativei/controller/LeadController.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            Lead lead = leadService.getById(id);
+            if(lead == null){
+                logger.error("unable to find lead for id: " + id);
+                return new ResponseEntity<>(ResponseObject.getResponse(ExceptionType.NO_DATA_FOUND.getMessage(),
+                        ExceptionType.NO_DATA_FOUND.getCode()), HttpStatus.NOT_FOUND);
+            }
+            LeadVO leadVO = new LeadVO(lead);
+            return new ResponseEntity<>(ResponseObject.getResponse(leadVO), HttpStatus.OK);
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
```
Comment (aayushKhandpur) [inline src/main/java/creativei/controller/LeadController.java]: Avoid logging twice.

**Comment by manojkrsaini09** [comment]
lead edit functionality added

**Pull request merged**